File: src/pub.cpp
#include "ros/ros.h"
#include "performance_tests/SuperAwesome.h"
#include "std_msgs/Float32.h"

#include <sstream>

using namespace std;

double frequency = 100; //25000 for CPP Pub - CPP Sub pair;
int frequencyPub = 0;

void frequencyPubCallback(const ros::TimerEvent &event) {
  frequencyPub = 1;
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "sender");
  ros::NodeHandle n;
  ros::Publisher pub = n.advertise<performance_tests::SuperAwesome>("test", 1);
  ros::Publisher freqpub = n.advertise<std_msgs::Float32>("testpubfreq", 1);
  ros::Timer freqChangeTmr = n.createTimer(ros::Duration(5), &frequencyPubCallback);

  ros::Rate loop_rate(frequency);
  performance_tests::SuperAwesome msg;
  std_msgs::Float32 pubFreq;

  ROS_INFO("pubf = %f", frequency);

  while (ros::ok())
  {
    std::stringstream ss;
    ss << ros::Time::now();
    msg.data = ss.str();
    pubFreq.data = (double)frequency;

//    ROS_INFO("%s", msg.data.c_str());
    pub.publish(msg);

    if(frequencyPub == 1)
    {
      freqpub.publish(pubFreq);
      ROS_INFO("pubf = %f", pubFreq.data);
      frequencyPub = 0;
    }
    ros::spinOnce();

    loop_rate.sleep();
  }

  return 0;
}

File: scripts/pub.py
#!/usr/bin/env python

import rospy
from performance_tests.msg import SuperAwesome
from std_msgs.msg import Float32

freqpublish = 0

def sender():
    pub = rospy.Publisher('test', SuperAwesome, queue_size=1)
    freqpub = rospy.Publisher('testpubfreq', Float32, queue_size=1);
    rospy.init_node('sender', anonymous=True)
    rospy.Timer(rospy.Duration(5), freqpub_callback)
    frequency = 100
    rate = rospy.Rate(frequency)
    msg = SuperAwesome()
    pubFreq = Float32()
    while not rospy.is_shutdown():
        msg.data = "%s" % rospy.Time.now()
        pubFreq.data = frequency
        pub.publish(msg)
	if freqpublish == 1:
           freqpub.publish(pubFreq)
           print 'pubf = ' + str(pubFreq.data)
        rate.sleep()

def freqpub_callback(event):
    freqpublish = 1
    print 'Timer called at ' + str(event.current_real)
   
if __name__ == '__main__':
    try:
        sender() 
    except rospy.ROSInterruptException:
        pass
